### DISCUSSION TITLE
Revoking "Pending" Celery Task Question

### DISCUSSION THREAD

**Discussion by formlogic-robert** [celery/kombu#2325] [Q&A]
Is there a way to revoke a pending celery task where the transport is kombu and broker is redis?
Currently the revoke won't happen if other tasks are ahead of the task being revokved, it will wait until it's that tasks turn to actually revoke it. Is there a way around this?

**Comment by TayfurYldz** [comment]
A Celery revoke is not a Redis operation that removes a particular serialized task from the middle of the broker list. Workers learn that a task ID is revoked and discard it when they encounter/reserve it. That explains why a task sitting behind other queued messages can remain visibly pending until consumption reaches it.

If the requirement is "delete this arbitrary task from the Redis queue immediately", normal revoke doesn't provide that guarantee. Manipulating the broker's list directly is risky because you'd be depending on Celery/Kombu's serialized message representation and queue bookkeeping.

For jobs that must be cancellable before they start, a safer application design is to keep cancellation state keyed by task/job ID (database/Redis key) and have the task check that state as soon as it is received, while still using revoke for the worker-side discard mechanism. If queue ordering itself is the problem, separating workloads into queues/priorities can also prevent a long backlog from delaying when cancellation is observed.
